### PR TITLE
Implement script parameters

### PR DESCRIPTION
**Pull request by benjaminkvm**
Work towards https://github.com/dapperlabs/flow-go/issues/3068.

## Description

Provide support for script parameters, with corresponding test cases.

### CODE REVIEW

File: runtime/errors.go [hunk 1]
```diff
@@ -78,6 +78,36 @@ func (e InvalidTransactionCountError) Error() string {
...
+// MissingEntryPointError
+
+type MissingEntryPointError struct {
+	Expected string
+	Err      error
+}
+
+func (e *MissingEntryPointError) Unwrap() error {
+	return e.Err
+}
```
Comment (turbolent) [inline runtime/errors.go:90]: What is the purpose of the wrapped error here? It looks like it's only ever created without an inner error: https://github.com/onflow/cadence/pull/151/files#diff-6681b213d55a5c65c5d90b211107fd52R136

Maybe just remove the `Err` field and `Unwrap` method.

File: runtime/errors.go [hunk 2]
```diff
@@ -78,6 +78,36 @@ func (e InvalidTransactionCountError) Error() string {
 	)
 }
 
+// MissingEntryPointError
```
Comment (turbolent) [inline runtime/errors.go:81]: Thanks for making this a proper error 👍

File: runtime/errors.go [hunk 3]
```diff
@@ -78,6 +78,36 @@ func (e InvalidTransactionCountError) Error() string {
...
+// InvalidEntryPointError
+
+type InvalidEntryPointTypeError struct {
+	Type sema.Type
+	Err  error
+}
+
+func (e *InvalidEntryPointTypeError) Unwrap() error {
+	return e.Err
+}
```
Comment (turbolent) [inline runtime/errors.go:105]: Same here, maybe just remove the `Err` field and `Unwrap` method.

File: runtime/errors.go [hunk 4]
```diff
@@ -123,6 +153,21 @@ func (e *InvalidTransactionArgumentError) Error() string {
...
+// InvalidScriptArgumentError
+
+type InvalidScriptArgumentError struct {
+	Index int
+	Err   error
+}
+
+func (e *InvalidScriptArgumentError) Unwrap() error {
+	return e.Err
+}
```
Comment (turbolent) [inline runtime/errors.go:165]: Same here, maybe just remove the `Err` field and `Unwrap` method.

File: runtime/runtime.go [hunk 5]
```diff
@@ -161,6 +167,53 @@ func (r *interpreterRuntime) ExecuteScript(
...
+					Index: i,
+					Err: &InvalidTypeAssignmentError{
+						Value: arg,
+						Type:  parameterType,
+					},
+				}
+			}
+
+			argumentValues[i] = arg
+		}
```
Comment (turbolent) [inline runtime/runtime.go:211]: This is the same code as https://github.com/onflow/cadence/blob/a968f040503e86274f7ac9c522110ce66e99b110/runtime/runtime.go#L349-L386, which is correct, but maybe refactor both into a common function to ensure the logic stays the same for both entry points (transactions and scripts)

**Comment by benjaminkvm** [comment]
Still have the `transaction/scriptExecutionFunction` fyi though much smaller as they now call `validateArgumentParameters` to remove duplicate code, let me know if you would prefer those methods as closures in their respective methods.

Had to change the error name to something that would work for both scripts and transactions, let me know what you think!

File: runtime/errors.go [hunk 6]
```diff
@@ -78,14 +78,34 @@ func (e InvalidTransactionCountError) Error() string {
...
-type InvalidTransactionParameterCountError struct {
+type InvalidEntryPointParameterCountError struct {
 	Expected int
 	Actual   int
 }
 
-func (e InvalidTransactionParameterCountError) Error() string {
+func (e InvalidEntryPointParameterCountError) Error() string {
 	return fmt.Sprintf(
 		"parameter count mismatch for transaction: expected %d, got %d",
```
Comment (psiemens) [inline runtime/errors.go:110]:
```suggestion
		"parameter count mismatch for entry point: expected %d, got %d",
```

File: runtime/runtime.go [hunk 7]
```diff
@@ -130,22 +131,27 @@ func (r *interpreterRuntime) ExecuteScript(
 		return nil, newError(err)
 	}
 
-	_, ok := checker.GlobalValues["main"]
+	ep, ok := checker.GlobalValues["main"]
 	if !ok {
-		// TODO: error because no main?
-		return nil, nil
+		return nil, &MissingEntryPointError{Expected: "main"}
```
Comment (psiemens) [inline runtime/runtime.go:136]: 💯

File: runtime/runtime_test.go [hunk 8]
```diff
@@ -751,6 +751,276 @@ func TestRuntimeTransactionWithArguments(t *testing.T) {
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.label, func(t *testing.T) {
+			rt := NewInterpreterRuntime()
+
+			var loggedMessages []string
+
+			runtimeInterface := &testRuntimeInterface{
+				// getSigningAccounts: func() []Address { return nil },
```
Comment (psiemens) [inline runtime/runtime_test.go:993]:
```suggestion
```

File: runtime/runtime_test.go [hunk 9]
```diff
@@ -751,6 +751,276 @@ func TestRuntimeTransactionWithArguments(t *testing.T) {
...
+
+			if tt.check != nil {
+				tt.check(t, err)
+			} else {
+				if !assert.NoError(t, err) {
+					for err := err; err != nil; err = errors.Unwrap(err) {
+						t.Log(err)
+					}
+				}
+				fmt.Println(loggedMessages)
```
Comment (psiemens) [inline runtime/runtime_test.go:1017]: Best to use `t.Log` if this is still needed

**Comment by psiemens** [comment]
@turbolent This PR is blocked on a code owner approval, but I'm going to merge as an admin in order to unblock the remaining work in `flow-go` and `flow-emulator` this sprint. Let us know if anything else comes up and we can open up a quick follow-up PR.

**Comment by psiemens** [comment]
Never mind, didn't need to break protocol!

**Pull request merged**